Refactors tracing handler to use HttpServerParser

hyleung opened the pull request:
This commit introduces an implementaton of `HttpServerParser` and
refactors the server tracing handler such that span name customization
is done via the `HttpServerParser` instead of directly in the tracing
handler.

In order to make this work, needed to add a wrapper around the standard
Ratpack `Response` so that things like the `PathBinding` and original
`ServerRequest` are available to the `HttpServerParser`.

Closes #51

hyleung committed:
Refactors tracing handler to use HttpServerParser

This commit introduces an implementaton of HttpServerParser and
refactors the server tracing handler such that span name customization
is done via the HttpServerParser instead of directly in the tracing
handler.

In order to make this work, needed to add a wrapper around the standard
Ratpack Response so that things like the PathBinding and original
ServerRequest are available to the HttpServerParser.

Closes #51

codefromthecrypt commented:
Looks good. FYI I will be raising something soon for "http.template", which is a tag that is uniform across templated requests, though not guaranteed to be a URI template.

hyleung commented:
@adriancole thanks for taking a look! I'm wondering if I can get rid of all the explicit Span start/finish stuff that's in `DefaultServerTracingHandler`. As far as I can tell, that should all be taken care of by the `HttpServerHandler`. 

https://github.com/hyleung/ratpack-zipkin/blob/ae1a8367e0deddbcc5df9181e039937b097d7886/src/main/java/ratpack/zipkin/internal/DefaultServerTracingHandler.java#L46-L55

hyleung committed:
Remove Span start/finish from tracing handler

Not needed anymore because Brave's HttpServerHandler takes care of all
this now.

hyleung commented on src/main/java/ratpack/zipkin/internal/DefaultServerTracingHandler.java:
```diff
@@ -28,32 +30,29 @@
 
+    //create a scope so that we can @Inject a SpanCustomizer into
+    //handlers and add tags, annotations, etc.
+    final Tracer.SpanInScope scope = tracing.tracer().withSpanInScope(span);
```
I don't actually know whether this is "a thing" - i.e. wanting to be able `@Inject` a `SpanCustomizer` into a handler and add tags, etc. I guess it might be useful if it's something that you don't want to implement as a general "policy" using something like a custom `HttpServerParser` ¯\\_(ツ)_/¯

codefromthecrypt replied:
maybe harder hitting comment could be.. "place the scan in scope so that downstream code, such as logging can see the ids"

codefromthecrypt commented on src/main/java/ratpack/zipkin/internal/DefaultServerTracingHandler.java:
```diff
@@ -28,32 +30,29 @@
     ctx.getResponse().beforeSend(response -> {
-      span.name(spanNameProvider.spanName(request, Optional.ofNullable(ctx.getPathBinding())));
-      handler.handleSend(response, null, span);
-      span.finish();
```
I think you asked.. yes, removing explicit start/finish was a good choice!
